README.md
# go-cos

腾讯云对象存储服务 COS(Cloud Object Storage) Go SDK（API 版本：V5 版本的 XML API）。

[![Build Status](https://img.shields.io/travis/mozillazg/go-cos/master.svg)](https://travis-ci.org/mozillazg/go-cos)
[![Coverage Status](https://img.shields.io/codecov/c/github/mozillazg/go-cos/master.svg)](https://codecov.io/gh/mozillazg/go-cos)
[![Go Report Card](https://goreportcard.com/badge/github.com/mozillazg/go-cos)](https://goreportcard.com/report/github.com/mozillazg/go-cos)
[![GoDoc](https://godoc.org/github.com/mozillazg/go-cos?status.svg)](https://godoc.org/github.com/mozillazg/go-cos)

## Install

`go get -u github.com/mozillazg/go-cos`

推荐使用 [go mod](https://github.com/golang/go/wiki/Modules) 之类的技术指定使用的 go-cos 包版本号。


## Usage

```go
package main

import (
	"context"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"

	"github.com/mozillazg/go-cos"
)

func main() {
	b, _ := cos.NewBaseURL("https://<bucket>-<appid>.cos.<region>.myqcloud.com")
	c := cos.NewClient(b, &http.Client{
		Transport: &cos.AuthorizationTransport{
			SecretID:  os.Getenv("COS_SECRETID"),
			SecretKey: os.Getenv("COS_SECRETKEY"),
		},
	})

	name := "test/hello.txt"
	resp, err := c.Object.Get(context.Background(), name, nil)
	if err != nil {
		panic(err)
	}
	defer resp.Body.Close()
	bs, _ := ioutil.ReadAll(resp.Body)
	fmt.Printf("%s\n", string(bs))
}
```

备注：

* SDK 不会自动设置超时时间，用户根据需要设置合适的超时时间（比如，设置 `http.Client` 的 `Timeout` 字段或者
  `Transport` 字段之类的）或在需要时实现所需的超时机制（比如，通过 `context` 包实现）。
* 所有的 API 在 [_example](./_example/) 目录下都有对应的使用示例（示例程序中用到的 `debug` 包只是调试用的不是必需的依赖）。

## TODO

Service API:

* [x] Get Service（使用示例：[service/get.go](./_example/service/get.go)）

Bucket API:

* [x] **Get Bucket**（搜索文件，使用示例：[bucket/get.go](./_example/bucket/get.go)）
* [x] Get Bucket ACL（使用示例：[bucket/getACL.go](./_example/bucket/getACL.go)）
* [x] Get Bucket CORS（使用示例：[bucket/getCORS.go](./_example/bucket/getCORS.go)）
* [x] Get Bucket Location（使用示例：[bucket/getLocation.go](./_example/bucket/getLocation.go)）
* [x] Get Buket Lifecycle（使用示例：[bucket/getLifecycle.go](./_example/bucket/getLifecycle.go)）
* [x] Get Bucket Tagging（使用示例：[bucket/getTagging.go](./_example/bucket/getTagging.go)）
* [ ] Get Bucket policy
* [x] Put Bucket（创建 bucket，使用示例：[bucket/put.go](./_example/bucket/put.go)）
* [x] Put Bucket ACL（使用示例：[bucket/putACL.go](./_example/bucket/putACL.go)）
* [x] Put Bucket CORS（使用示例：[bucket/putCORS.go](./_example/bucket/putCORS.go)）
* [x] Put Bucket Lifecycle（使用示例：[bucket/putLifecycle.go](./_example/bucket/putLifecycle.go)）
* [x] Put Bucket Tagging（使用示例：[bucket/putTagging.go](./_example/bucket/putTagging.go)）
* [ ] Put Bucket policy
* [x] Delete Bucket（删除 bucket，使用示例：[bucket/delete.go](./_example/bucket/delete.go)）
* [x] Delete Bucket CORS（使用示例：[bucket/deleteCORS.go](./_example/bucket/deleteCORS.go)）
* [x] Delete Bucket Lifecycle（使用示例：[bucket/deleteLifecycle.go](./_example/bucket/deleteLifecycle.go)）
* [x] Delete Bucket Tagging（使用示例：[bucket/deleteTagging.go](./_example/bucket/deleteTagging.go)）
* [ ] Delete Bucket policy
* [x] Head Bucket（使用示例：[bucket/head.go](./_example/bucket/head.go)）
* [x] List Multipart Uploads（查询上传的分块，使用示例：[bucket/listMultipartUploads.go](./_example/bucket/listMultipartUploads.go)）

Object API:

* [x] **Append Object**（增量更新文件，使用示例：[object/append.go](./_example/object/append.go)）
* [x] **Get Object**（下载文件，使用示例：[object/get.go](./_example/object/get.go)）
* [x] Get Object ACL（使用示例：[object/getACL.go](./_example/object/getACL.go)）
* [x] **Put Object**（上传文件，使用示例：[object/put.go](./_example/object/put.go) or [object/uploadFile.go](./_example/object/uploadFile.go)）
* [x] Put Object ACL（使用示例：[object/putACL.go](./_example/object/putACL.go)）
* [x] Put Object Copy（使用示例：[object/copy.go](./_example/object/copy.go)）
* [x] **Delete Object**（删除文件，使用示例：[object/delete.go](./_example/object/delete.go)）
* [ ] [Post Object](https://cloud.tencent.com/document/product/436/14690)
* [ ] [Post Object restore](https://cloud.tencent.com/document/product/436/12633)
* [x] Delete Multiple Object（使用示例：[object/deleteMultiple.go](./_example/object/deleteMultiple.go)）
* [x] Head Object（使用示例：[object/head.go](./_example/object/head.go)）
* [x] Options Object（使用示例：[object/options.go](./_example/object/options.go)）
* [x] **Initiate Multipart Upload**（初始化分块上传，使用示例：[object/initiateMultipartUpload.go](./_example/object/initiateMultipartUpload.go)）
* [x] **Upload Part**（上传一个分块，使用示例：[object/uploadPart.go](./_example/object/uploadPart.go)）
* [ ] [Upload Part - Copy](https://cloud.tencent.com/document/product/436/8287)
* [x] **List Parts**（列出已上传的分块，使用示例：[object/listParts.go](./_example/object/listParts.go)）
* [x] **Complete Multipart Upload**（合并上传的分块，使用示例：[object/completeMultipartUpload.go](./_example/object/completeMultipartUpload.go)）
* [x] **Abort Multipart Upload**（取消分块上传，使用示例：[object/abortMultipartUpload.go](./_example/object/abortMultipartUpload.go)）

其他功能：

* [x] **生成预签名授权 URL**
    * [x] 通过预签名授权 URL 下载文件，示例：[object/getWithPresignedURL.go](./_example/object/getWithPresignedURL.go)
    * [x] 通过预签名授权 URL 上传文件，示例：[object/putWithPresignedURL.go](./_example/object/putWithPresignedURL.go)
* [x] **支持临时密钥**，示例: [object/sessionToken.go](./_example/object/sessionToken.go)
* [x] 支持使用使用第三方 http client 包或单元测试时 mock 方法调用结果，示例：[object/mock.go](./_example/object/mock.go)


CHANGELOG.md
# Changelog

## [0.13.0] (2019-xx-yy)

## 新增

* `AuthorizationTransport` 增加 `SessionToken` 字段，支持通过临时密钥请求 api，示例: [object/sessionToken.go](./_example/object/sessionToken.go) 。
* 新增 `c.Object.ListPartsWithOpt` 方法，解决 `c.Object.ListParts` 方法忘了支持参数的问题。
  * `ListPartsWithOpt(ctx context.Context, name, uploadID string, opt *ObjectListPartsOptions) (*ObjectListPartsResult, *Response, error)`
  * `ListParts(ctx context.Context, name, uploadID string) (*ObjectListPartsResult, *Response, error)`


## [0.12.0] (2019-04-27)

### 新增

* 支持使用使用第三方 http client 包或单元测试时 mock 方法调用结果，示例：[object/mock.go](./_example/object/mock.go)
  * 新增 `type Sender interface`
  * 新增 `type ResponseParser interface`
  * 新增 `type DefaultSender struct`
  * 新增 `type DefaultResponseParser struct`


## [0.11.1] (2019-04-14)

### Bugfix

* 修复当 url 或 headers 相关参数的值中包含空格或某几个特殊字符串时会出现服务端返回签名不匹配的问题。(via [#13])


## [0.11.0] (2018-12-08)

### 不兼容旧版的变更

* 根据 COS [官方文档](https://cloud.tencent.com/document/product/436/7751) 更新最新版本下
  各个 API 的参数和响应中包含的字段（新增了部分字段，废弃了一些字段）。

影响：

* 大部分用户无需修改任何代码，部分依赖了字段结构的高级用户需要更新代码。

### 新增

* `Response` 新增几个方法用于快速获取常用的 COS Header 的值:
  * `RequestID()`
  * `TraceID()`
  * `ObjectType()`
  * `StorageClass()`
  * `VersionID()`
  * `ServerSideEncryption()`
  * `MetaHeaders()`
* 新增了几个常量用于判断部分固定值

### 文档

* 上传文件操作不再需要在特定情况下强制指定 ContentLength 了（COS 服务端新功能）。
* 强调一下用户可以自己设置超时时间或通过 context 实现终止请求的功能。


## [0.10.0] (2018-11-03)

### 变更

* 当上传文件相关方法的 `r io.Reader` 参数是个 `io.ReadCloser` 时不会再自动调用 `r.Close()` ，
  用户需要自行选择合适的时机去调用 `r.Close()` 方法对 r 进行资源回收。(via [#7] Thanks [@jojohappy])


## [0.9.0] (2018-08-04)

### 新增

* 新增 `c.Object.PresignedURL` 用于获取预签名授权 URL。
  可用于无需知道 SecretID 和 SecretKey 就可以上传和下载文件。
* 上传和下载 Object 的功能支持指定预签名授权 URL。

详见 PR 以及使用示例：

* https://github.com/mozillazg/go-cos/pull/5
* 通过预签名授权 URL 下载文件，示例：[object/getWithPresignedURL.go](./_example/object/getWithPresignedURL.go)
* 通过预签名授权 URL 上传文件，示例：[object/putWithPresignedURL.go](./_example/object/putWithPresignedURL.go)


## [0.8.0] (2018-05-26)

### 新增

* 新增 `func NewBaseURL(bucketURL string) (u *BaseURL, err error)` (via [91f7759])

### 不兼容旧版的变更

* `NewBucketURL` 函数使用新的 URL 域名规则。(via [7dcd701])

影响：

* 如果有使用 `NewBucketURL` 函数生成 bucketURL 的话，使用时需要使用新的 Region 名称，
详见 https://cloud.tencent.com/document/product/436/6224 ，未使用 `NewBucketURL` 函数不受影响


## [0.7.0] (2017-12-23)

### 新增

* 支持新增的 Put Object Copy API
* 新增 `github.com/mozillazg/go-cos/debug`，目前只包含 `DebugRequestTransport`


## [0.6.0] (2017-07-09)

### 新增

* 增加说明在某些情况下 ObjectPutHeaderOptions.ContentLength 必须要指定
* 增加 ObjectUploadPartOptions.ContentLength


## [0.5.0] (2017-06-28)

### 修复

* 修复 ACL 相关 API 突然失效的问题.
  (因为 COS ACL 相关 API 的 request 和 response xml body 的结构发生了变化)

### 删除

* 删除调试用的 DebugRequestTransport(把它移动到 examples/ 中)


## [0.4.0] (2017-06-24)

### 新增

* 增加 AuthorizationTransport 辅助添加认证信息

### 修改

* 去掉 API 中的 authTime 参数，默认不再自动添加 Authorization header
  改为通过自定义 client 的方式来添加认证信息


## [0.3.0] (2017-06-23)

### 新增

* 完成剩下的所有 API


## [0.2.0] (2017-06-10)

### 不兼容旧版的变更

* 调用 bucket 相关 API 时不再需要 bucket 参数, 把参数移到 service 中
* 把参数 signStartTime, signEndTime, keyStartTime, keyEndTime 合并为 authTime


## 0.1.0 (2017-06-10)

### 新增

* 完成 Service API
* 完成大部分 Bucket API(还剩一个 Put Bucket Lifecycle)


[0.13.0]: https://github.com/mozillazg/go-cos/compare/v0.12.0...v0.13.0
[0.12.0]: https://github.com/mozillazg/go-cos/compare/v0.11.1...v0.12.0
[0.11.1]: https://github.com/mozillazg/go-cos/compare/v0.11.0...v0.11.1
[0.11.0]: https://github.com/mozillazg/go-cos/compare/v0.10.0...v0.11.0
[0.10.0]: https://github.com/mozillazg/go-cos/compare/v0.9.0...v0.10.0
[0.9.0]: https://github.com/mozillazg/go-cos/compare/v0.8.0...v0.9.0
[0.8.0]: https://github.com/mozillazg/go-cos/compare/v0.7.0...v0.8.0
[0.7.0]: https://github.com/mozillazg/go-cos/compare/v0.6.0...v0.7.0
[0.6.0]: https://github.com/mozillazg/go-cos/compare/v0.5.0...v0.6.0
[0.5.0]: https://github.com/mozillazg/go-cos/compare/v0.4.0...v0.5.0
[0.4.0]: https://github.com/mozillazg/go-cos/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/mozillazg/go-cos/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/mozillazg/go-cos/compare/v0.1.0...v0.2.0

[91f7759]: https://github.com/mozillazg/go-cos/commit/91f7759958f9631e8997f47d30ae4044455fc971
[7dcd701]: https://github.com/mozillazg/go-cos/commit/7dcd701975f483d57525b292ab31d0f9a6c8866c
[#7]: https://github.com/mozillazg/go-cos/pull/7
[@jojohappy]: https://github.com/jojohappy
[#13]: https://github.com/mozillazg/go-cos/pull/13


cos.go
package cos

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/xml"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"net/url"
	"reflect"
	"strconv"
	"strings"
	"text/template"

	"github.com/google/go-querystring/query"
	"github.com/mozillazg/go-httpheader"
)

const (
	// Version ...
	Version               = "0.12.0"
	userAgent             = "go-cos/" + Version
	contentTypeXML        = "application/xml"
	defaultServiceBaseURL = "https://service.cos.myqcloud.com"
)

var bucketURLTemplate = template.Must(
	template.New("bucketURLFormat").Parse(
		"{{.Scheme}}://{{.BucketName}}-{{.AppID}}.cos.{{.Region}}.myqcloud.com",
	),
)

// BaseURL 访问各 API 所需的基础 URL
type BaseURL struct {
	// 访问 bucket, object 相关 API 的基础 URL（不包含 path 部分）
	// 比如：https://test-1253846586.cos.ap-beijing.myqcloud.com
	// 详见 https://cloud.tencent.com/document/product/436/6224
	BucketURL *url.URL
	// 访问 service API 的基础 URL（不包含 path 部分）
	// 比如：https://service.cos.myqcloud.com
	ServiceURL *url.URL
}

// NewBaseURL 生成 BaseURL
func NewBaseURL(bucketURL string) (u *BaseURL, err error) {
	bu, err := url.Parse(bucketURL)
	if err != nil {
		return
	}
	su, _ := url.Parse(defaultServiceBaseURL)
	u = &BaseURL{
		BucketURL:  bu,
		ServiceURL: su,
	}
	return
}

// NewBucketURL 生成 BaseURL 所需的 BucketURL
//
//   bucketName: bucket 名称
//   AppID: 应用 ID
//   Region: 区域代码，详见 https://cloud.tencent.com/document/product/436/6224
//   secure: 是否使用 https
func NewBucketURL(bucketName, appID, region string, secure bool) *url.URL {
	scheme := "https"
	if !secure {
		scheme = "http"
	}

	w := bytes.NewBuffer(nil)
	bucketURLTemplate.Execute(w, struct {
		Scheme     string
		BucketName string
		AppID      string
		Region     string
	}{
		scheme, bucketName, appID, region,
	})

	u, _ := url.Parse(w.String())
	return u
}

// A Client manages communication with the COS API.
type Client struct {
	// Sender 用于实际发送 HTTP 请求
	Sender Sender
	// ResponseParser 用于解析响应
	ResponseParser ResponseParser

	UserAgent string
	BaseURL   *BaseURL

	common service

	// Service 封装了 service 相关的 API
	Service *ServiceService
	// Bucket 封装了 bucket 相关的 API
	Bucket *BucketService
	// Object 封装了 object 相关的 API
	Object *ObjectService
}

type service struct {
	client *Client
}

// NewClient returns a new COS API client.
// 使用 DefaultSender 作为 Sender，DefaultResponseParser 作为 ResponseParser
func NewClient(uri *BaseURL, httpClient *http.Client) *Client {
	if httpClient == nil {
		httpClient = &http.Client{}
	}

	baseURL := &BaseURL{}
	if uri != nil {
		baseURL.BucketURL = uri.BucketURL
		baseURL.ServiceURL = uri.ServiceURL
	}
	if baseURL.ServiceURL == nil {
		baseURL.ServiceURL, _ = url.Parse(defaultServiceBaseURL)
	}

	c := &Client{
		Sender:         &DefaultSender{httpClient},
		ResponseParser: &DefaultResponseParser{},
		UserAgent:      userAgent,
		BaseURL:        baseURL,
	}
	c.common.client = c
	c.Service = (*ServiceService)(&c.common)
	c.Bucket = (*BucketService)(&c.common)
	c.Object = (*ObjectService)(&c.common)
	return c
}

func (c *Client) newRequest(ctx context.Context, opt *sendOptions) (req *http.Request, err error) {
	baseURL := opt.baseURL
	uri := opt.uri
	method := opt.method
	body := opt.body
	optQuery := opt.optQuery
	optHeader := opt.optHeader

	uri, err = addURLOptions(uri, optQuery)
	if err != nil {
		return
	}
	u, _ := url.Parse(uri)
	urlStr := baseURL.ResolveReference(u).String()

	var reader io.Reader
	contentType := ""
	contentMD5 := ""
	xsha1 := ""
	if body != nil {
		// 上传文件
		if r, ok := body.(io.Reader); ok {
			reader = r
		} else {
			b, err := xml.Marshal(body)
			if err != nil {
				return nil, err
			}
			contentType = contentTypeXML
			reader = bytes.NewReader(b)
			contentMD5 = base64.StdEncoding.EncodeToString(calMD5Digest(b))
			// xsha1 = base64.StdEncoding.EncodeToString(calSHA1Digest(b))
		}
	} else {
		contentType = contentTypeXML
	}

	req, err = http.NewRequest(method, urlStr, reader)
	if err != nil {
		return
	}

	req.Header, err = addHeaderOptions(req.Header, optHeader)
	if err != nil {
		return
	}
	if v := req.Header.Get("Content-Length"); req.ContentLength == 0 && v != "" && v != "0" {
		req.ContentLength, _ = strconv.ParseInt(v, 10, 64)
		req.Body = ioutil.NopCloser(reader)
	}

	if contentMD5 != "" {
		req.Header["Content-MD5"] = []string{contentMD5}
	}
	if xsha1 != "" {
		req.Header.Set("x-cos-sha1", xsha1)
	}
	if c.UserAgent != "" {
		req.Header.Set("User-Agent", c.UserAgent)
	}
	if req.Header.Get("Content-Type") == "" && contentType != "" {
		req.Header.Set("Content-Type", contentType)
	}
	return
}

func (c *Client) doAPI(ctx context.Context, caller Caller, req *http.Request, result interface{}, closeBody bool) (*Response, error) {
	req = req.WithContext(ctx)
	resp, err := c.Sender.Send(ctx, caller, req)
	if err != nil {
		return nil, err
	}

	defer func() {
		if closeBody {
			// Close the body to let the Transport reuse the connection
			io.Copy(ioutil.Discard, resp.Body)
			resp.Body.Close()
		}
	}()

	return c.ResponseParser.ParseResponse(ctx, caller, resp, result)
}

type sendOptions struct {
	// 基础 URL
	baseURL *url.URL
	// URL 中除基础 URL 外的剩余部分
	uri string
	// 请求方法
	method string

	body interface{}
	// url 查询参数
	optQuery interface{}
	// http header 参数
	optHeader interface{}
	// 用 result 反序列化 resp.Body
	result interface{}
	// 是否禁用自动调用 resp.Body.Close()
	// 自动调用 Close() 是为了能够重用连接
	disableCloseBody bool

	caller Caller
}

func (c *Client) send(ctx context.Context, opt *sendOptions) (resp *Response, err error) {
	req, err := c.newRequest(ctx, opt)
	if err != nil {
		return
	}

	resp, err = c.doAPI(ctx, opt.caller, req, opt.result, !opt.disableCloseBody)
	if err != nil {
		return
	}
	return
}

// addURLOptions adds the parameters in opt as URL query parameters to s. opt
// must be a struct whose fields may contain "url" tags.
func addURLOptions(s string, opt interface{}) (string, error) {
	v := reflect.ValueOf(opt)
	if v.Kind() == reflect.Ptr && v.IsNil() {
		return s, nil
	}

	u, err := url.Parse(s)
	if err != nil {
		return s, err
	}

	qs, err := query.Values(opt)
	if err != nil {
		return s, err
	}

	// 保留原有的参数，并且放在前面。因为 cos 的 url 路由是以第一个参数作为路由的
	// e.g. /?uploads
	q := u.RawQuery
	rq := qs.Encode()
	if q != "" {
		if rq != "" {
			u.RawQuery = fmt.Sprintf("%s&%s", q, qs.Encode())
		}
	} else {
		u.RawQuery = rq
	}
	return u.String(), nil
}

// addHeaderOptions adds the parameters in opt as Header fields to req. opt
// must be a struct whose fields may contain "header" tags.
func addHeaderOptions(header http.Header, opt interface{}) (http.Header, error) {
	v := reflect.ValueOf(opt)
	if v.Kind() == reflect.Ptr && v.IsNil() {
		return header, nil
	}

	h, err := httpheader.Header(opt)
	if err != nil {
		return nil, err
	}

	for key, values := range h {
		for _, value := range values {
			header.Add(key, value)
		}
	}
	return header, nil
}

// Owner ...
type Owner struct {
	UIN         string `xml:"uin,omitempty"`
	ID          string `xml:",omitempty"`
	DisplayName string `xml:",omitempty"`
}

// Initiator ...
type Initiator Owner

// Response API 响应
type Response struct {
	*http.Response
}

func newResponse(resp *http.Response) *Response {
	return &Response{
		Response: resp,
	}
}

var (
	xCosRequestID            = "x-cos-request-id"
	xCosTraceID              = "x-cos-trace-id"
	xCosObjectType           = "x-cos-object-type"
	xCosStorageClass         = "x-cos-storage-class"
	xCosVersionID            = "x-cos-version-id"
	xCosServerSideEncryption = "x-cos-server-side-encryption"
	xCosMetaPrefix           = "x-cos-meta-"
)

// RequestID 每次请求发送时，服务端将会自动为请求生成一个ID。
func (resp *Response) RequestID() string {
	return resp.Header.Get(xCosRequestID)
}

// TraceID 每次请求出错时，服务端将会自动为这个错误生成一个ID。
func (resp *Response) TraceID() string {
	return resp.Header.Get(xCosTraceID)
}

// ObjectType 用来表示 Object 是否可以被追加上传，枚举值：normal 或者 appendable
func (resp *Response) ObjectType() string {
	return resp.Header.Get(xCosObjectType)
}

// StorageClass Object 的存储级别，枚举值：STANDARD，STANDARD_IA
func (resp *Response) StorageClass() string {
	return resp.Header.Get(xCosStorageClass)
}

// VersionID 如果检索到的对象具有唯一的版本ID，则返回版本ID。
func (resp *Response) VersionID() string {
	return resp.Header.Get(xCosVersionID)
}

// ServerSideEncryption 如果通过 COS 管理的服务端加密来存储对象，响应将包含此头部和所使用的加密算法的值，AES256。
func (resp *Response) ServerSideEncryption() string {
	return resp.Header.Get(xCosServerSideEncryption)
}

// MetaHeaders 用户自定义的元数据
func (resp *Response) MetaHeaders() http.Header {
	h := http.Header{}
	for k := range resp.Header {
		if !strings.HasPrefix(strings.ToLower(k), xCosMetaPrefix) {
			continue
		}
		for _, v := range resp.Header[k] {
			h.Add(k, v)
		}
	}
	return h
}

// ACLHeaderOptions ...
type ACLHeaderOptions struct {
	// 定义 Object 的 acl 属性。有效值：private，public-read-write，public-read；默认值：private
	XCosACL string `header:"x-cos-acl,omitempty" url:"-" xml:"-"`
	// 	赋予被授权者读的权限。格式：id="[OwnerUin]"
	XCosGrantRead string `header:"x-cos-grant-read,omitempty" url:"-" xml:"-"`
	// 赋予被授权者写的权限。格式：id="[OwnerUin]"
	XCosGrantWrite string `header:"x-cos-grant-write,omitempty" url:"-" xml:"-"`
	// 赋予被授权者所有的权限。格式：id="[OwnerUin]"
	XCosGrantFullControl string `header:"x-cos-grant-full-control,omitempty" url:"-" xml:"-"`
}

// ACLGrantee ...
type ACLGrantee struct {
	Type        string `xml:"type,attr"`
	UIN         string `xml:"uin,omitempty"`
	ID          string `xml:",omitempty"`
	DisplayName string `xml:",omitempty"`
	SubAccount  string `xml:"Subaccount,omitempty"`
}

// ACLGrant ...
type ACLGrant struct {
	Grantee *ACLGrantee
	// 指明授予被授权者的权限信息，枚举值：READ，WRITE，FULL_CONTROL
	Permission string
}

// ACLXml ...
//
// https://cloud.tencent.com/document/product/436/7733
type ACLXml struct {
	XMLName           xml.Name `xml:"AccessControlPolicy"`
	Owner             *Owner
	AccessControlList []ACLGrant `xml:"AccessControlList>Grant,omitempty"`
}

const (
	// StorageClassStandard Object 的存储级别: STANDARD
	StorageClassStandard string = "STANDARD"
	// StorageClassStandardTA Object 的存储级别: STANDARD_IA
	StorageClassStandardTA string = "STANDARD_IA"
	// StorageClassArchive Object 的存储级别: ARCHIVE
	StorageClassArchive string = "ARCHIVE"

	// ObjectTypeAppendable : appendable
	ObjectTypeAppendable string = "appendable"
	// ObjectTypeNormal : normal
	ObjectTypeNormal string = "normal"

	// ServerSideEncryptionAES256 服务端加密算法: AES256
	ServerSideEncryptionAES256 string = "AES256"

	// PermissionRead 权限值: READ
	PermissionRead string = "READ"
	// PermissionWrite 权限值: WRITE
	PermissionWrite string = "WRITE"
	// PermissionFullControl  权限值: FULL_CONTROL
	PermissionFullControl string = "FULL_CONTROL"
)
